### ISSUE TITLE
Ship DevTools with dart2wasm

### ISSUE THREAD

**Issue by kenzieschmoll** [flutter/devtools#7856]
The DevTools codebase is prepared for building DevTools with dart2wasm thanks to the unblocking work [here](https://github.com/flutter/devtools/issues/6606).

There are still a few requirements that we need to meet before moving this effort forward:

- [ ] Land the [bootstrapping changes](https://github.com/flutter/devtools/pull/7599) to DevTools in a way that does not break our g3 roll. 
  - CC @eyebrowsoffire @xster we will need support from the Fl@G team to fully support the new bootstrapping syntax. If adding full support is infeasible right now, we may have to workaround this by diverging the g3 bootstrapping code from the external. This is not ideal since this will cause a continued maintenance burden and potential regressions in g3, given we will not be testing externally with the same bootstrapping used in g3.
- [ ] Ensure that the VS Code embedded web views support WasmGC (see [background information on Chromium and V8](https://docs.flutter.dev/platform-integration/web/wasm#background))
  - @DanTup can you confirm?
- [ ] Ensure that the IntelliJ embedded web views (JXBrowser) support WasmGC (see [background information on Chromium and V8](https://docs.flutter.dev/platform-integration/web/wasm#background))
  - @helin24 can you confirm?
- [ ] Add support for opt in / opt out behavior. See these [web bootstrapping docs](https://docs.flutter.dev/platform-integration/web/bootstrapping#example-customizing-flutter-configuration-based-on-url-query-parameters) for an example of customizing configuration based on query parameters.

Compare the dart2wasm DevTools build to dart2js and file any issues that come up:
- TODO

**Comment by kevmoo** [comment]
@kenzieschmoll – keep in mind, we can use query param magic to force loading certain versions. So if there are contexts where wasm won't work, we can just disable them. Or we can make loading the wasm version opt-in for testing, etc.

**Comment by kevmoo** [comment]
Ha! just read your last bullet. 🤦

**Comment by DanTup** [comment]
> Ensure that the VS Code embedded web views support WasmGC (see [background information on Chromium and V8](https://docs.flutter.dev/platform-integration/web/wasm#background))

The current version of VS Code uses Chromium v120 which sounds like it should support this. However, I don't know if there are things in Electron/VS Code themselves that might impact the user of this. @kenzieschmoll what's the easiest way for me to build DevTools using WASM (can I do it through `devtools_tool serve`? hacking scripts locally to make this work would be fine for testing).

And when it is running, how do I confirm it is using the wasm version and didn't fall back to js?

**Comment by kenzieschmoll** [comment]
> what's the easiest way for me to build DevTools using WASM (can I do it through `devtools_tool serve`? hacking scripts locally to make this work would be fine for testing).
> 
> And when it is running, how do I confirm it is using the wasm version and didn't fall back to js?

Patch the changes from https://github.com/flutter/devtools/pull/7867 and https://dart-review.googlesource.com/c/sdk/+/369100. Then run `devtools_tool serve --wasm`.

If the wasm changes are picked up, you should see `main.dart.wasm`, `main.dart.mjs`, and `canvaskit/skwasm.js`, `canvaskit/skwasm.wasm` in the loaded sources. Here is what I see locally.
![Screenshot 2024-05-31 at 8 45 45 AM](https://github.com/flutter/devtools/assets/43759233/8fde943f-4380-47f6-8d4b-b42970bfbc8e)

**Comment by helin24** [comment]
IntelliJ uses JxBrowser version 7.38.2 currently, which includes Chromium 124.0.6367.92 and should support WasmGC.

It's hard for me to tell what version of JCEF IntelliJ uses when that option is chosen by user setting, but JCEF is a few versions behind the CEF library, which currently is using Chromium 125 (https://cef-builds.spotifycdn.com/index.html).

In summary I think either implementation of embedded browser will be fine for wasm.

**Comment by xster** [comment]
@yjbanov the sequencing will likely be that we'll work on shipping devtools in wasm in google after we decide to ship flutter web in wasm in google3

**Comment by DanTup** [comment]
@kenzieschmoll 

> Patch the changes from https://github.com/flutter/devtools/pull/7867 and https://dart-review.googlesource.com/c/sdk/+/369100. Then run devtools_tool serve --wasm.

The first has already landed but I patched in the second. I added "--wasm" to my `dart.customDevTools` setting. Inside VS Code, it seems to have loaded the JS version:

![image](https://github.com/flutter/devtools/assets/1078012/8a1f86b2-a2cf-4507-a965-cb6f301ef1dd)

However if I open in an external browser it loads the wasm version, however it throws and the app doesn't load:

![image](https://github.com/flutter/devtools/assets/1078012/843604e3-8353-4446-a64d-0bd25941aadd)

I presume the first issue is that whatever is being used to detect whether wasm is available is failing. @kevmoo is there an easy way for me to review (or step through) these conditions? Looking through `flutter_bootstrap.js` I found a function that looked like it was doing that which runs:

```
let o = [0, 97, 115, 109, 1, 0, 0, 0, 1, 5, 1, 95, 1, 120, 0];
return WebAssembly.validate(new Uint8Array(o))
```

However if I run that myself in the console, it appears to return `true`, so I think there may be other conditions.

![image](https://github.com/flutter/devtools/assets/1078012/62b0f7a3-7b8f-48fd-84b1-dd23704afaac)

**Comment by DanTup** [comment]
Seems like the issue is that `window.crossOriginIsolated` is `false`

![image](https://github.com/flutter/devtools/assets/1078012/2c867693-43ad-40af-8464-4504386c90a9)

The [docs here](https://developer.mozilla.org/en-US/docs/Web/API/Window/crossOriginIsolated) confirm the headers in the CL above are required, but I see those are also set:

![image](https://github.com/flutter/devtools/assets/1078012/4172d811-367b-4384-9a52-74e4da9c5583)

It's not clear to me why this value is `false`. There's a similar issue at https://github.com/microsoft/vscode/issues/183765 from last year about this being the case on Github.dev and a link to a private repo with a fix, which I suspect is just setting the same headers.

**Comment by DanTup** [comment]
I tracked this back to https://github.com/microsoft/vscode/issues/186614. It seems cross-origin-isolatation is currently disabled by VS Code. When it was enabled the were performance issues that resulted in it being reverted.

I tried running `code --enable-coi` to force this enabled for testing, however the iframe with DevTools in it won't load at all:

![image](https://github.com/flutter/devtools/assets/1078012/2bcf4c86-f591-4cf1-89ba-d4f493237b81)

Which seems to be because another header is not present:

![image](https://github.com/flutter/devtools/assets/1078012/c8ae3d08-71c8-4dd9-9db6-a67d18961d84)

I don't know the implications of adding that header.

**Comment by eyebrowsoffire** [comment]
Yes, if you're loading the app in an iframe, you will need to set a CORP header for the child document if you want cross origin isolation (which is needed for the multi-threaded rendering that the skwasm renderer does). I don't know exactly how the embedding of this works with VS Code. I think `Cross-Origin-Resource-Policy: cross-origin` should be fine for this use case, as long as all the assets that devtools uses are also served with `Cross-Origin-Resource-Policy: cross-origin`.

**Comment by DanTup** [comment]
Thanks! Adding that header on its own didn't work, but then I found I needed `allow="cross-origin-isolated"` on the iframe. With that, the flags are all true and the wasm version is used - and fails with:

![image](https://github.com/flutter/devtools/assets/1078012/f601e5c7-0321-4a41-8851-8c044ad73742)

This is the same error I posted above when running outside of VS Code. It's not clear to me if that's a DevTools or wasm bug?

**Comment by kenzieschmoll** [comment]
> This is the same error I posted above when running outside of VS Code. It's not clear to me if that's a DevTools or wasm bug?

I can't reproduce on Mac, so I am wondering if this is a Windows issue with wasm. CC @eyebrowsoffire

**Comment by eyebrowsoffire** [comment]
@DanTup Could you try building again with `--name-section` enabled so we get a more useful stack trace? `flutter build web --wasm --name-section`

**Comment by DanTup** [comment]
@eyebrowsoffire that flag didn't seem to work:

```
C:\Dev\Google\devtools\packages\devtools_app > C:\Dev\Google\devtools\tool\flutter-sdk\bin\flutter.bat build web --wasm --name-section --pwa-strategy=offline-first --no-tree-shake-icons
Could not find an option named "name-section".
```

I tried `--no-strip-wasm` (which I saw in https://github.com/dart-lang/sdk/commit/a98ce03e13c530f919e79512119271a659c35974) which initially gave the same error, but Ctrl+refresh resulted in:

```
Uncaught RuntimeError: illegal cast
    at _isFlutterGAEnabled inner (main.dart.wasm:0x38f4da)
    at _awaitHelperWithTypeCheck closure at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:97:16 (main.dart.wasm:0x1fdf8e)
    at closure wrapper at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:97:16 trampoline (main.dart.wasm:0x1fe083)
    at _rootRunUnary (main.dart.wasm:0x1febae)
    at _CustomZone.runUnary (main.dart.wasm:0x50f355)
    at _Future._propagateToListeners (main.dart.wasm:0x1fe77a)
    at _Future._completeWithValue (main.dart.wasm:0x1fedfc)
    at _Future._asyncCompleteWithValue closure at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:735:29 (main.dart.wasm:0x20fbd6)
    at closure wrapper at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:735:29 trampoline (main.dart.wasm:0x20fbec)
    at _rootRun (main.dart.wasm:0x1fd210)
```

**Comment by eyebrowsoffire** [comment]
Yes, sorry, I meant `--no-strip-wasm`, I got the `dart compile wasm` commands confused with the flutter tool's.

There are no casts in that function from the user code that I can see. I wonder if something is being inlined? You can also pass `-O0` and see if that disables some inlining and that might clarify what's going on here a bit.

**Comment by DanTup** [comment]
> There are no casts in that function from the user code that I can see

It's assigning `responseValue` (if not `null`) to `enabled`. `enabled` is a bool, `responseValue` is a string (and my IDE says `responseValue` is `dynamic`. It's not clear to me why this would behave different for wasm though.

Using `-O0` I get this (which does seem to line up):

```
main.dart.mjs:675 [zoneGuarded]: Type 'String' is not a subtype of type 'bool' in type cast
_1890 @ main.dart.mjs:675
$printToConsole @ main.dart.wasm:0x8156eb
$initDevToolsLogging closure at file:///C:/Dev/Google/devtools/packages/devtools_app/lib/src/shared/config_specific/logger/logger_helpers.dart:19:31 @ main.dart.wasm:0x8155d9
$closure wrapper at file:///C:/Dev/Google/devtools/packages/devtools_app/lib/src/shared/config_specific/logger/logger_helpers.dart:19:31 trampoline @ main.dart.wasm:0x815615
$_rootRunUnary @ main.dart.wasm:0x30582f
$_rootRunUnary tear-off trampoline @ main.dart.wasm:0x817ef3
$_CustomZone.runUnary @ main.dart.wasm:0x816525
$_CustomZone.runUnaryGuarded @ main.dart.wasm:0x8167fe
$_BufferingStreamSubscription._sendData @ main.dart.wasm:0x31e7c7
$_BufferingStreamSubscription._add @ main.dart.wasm:0x31ef90
$_SyncBroadcastStreamController._sendData @ main.dart.wasm:0x31e65f
$_BroadcastStreamController.add @ main.dart.wasm:0x31dad7
$Logger._publish @ main.dart.wasm:0x4721b5
$Logger.log @ main.dart.wasm:0x472005
$Logger.severe @ main.dart.wasm:0x471ec2
$reportError @ main.dart.wasm:0x641108
$setupErrorHandling closure at file:///C:/Dev/Google/devtools/packages/devtools_app/lib/src/framework/app_error_handling.dart:55:7 @ main.dart.wasm:0x8158f6
$closure wrapper at file:///C:/Dev/Google/devtools/packages/devtools_app/lib/src/framework/app_error_handling.dart:55:7 trampoline @ main.dart.wasm:0x815953
$_RootZone.runBinary @ main.dart.wasm:0x30572c
$runZonedGuarded closure at org-dartlang-sdk:///dart-sdk/lib/async/zone.dart:1795:45 @ main.dart.wasm:0x815b70
$closure wrapper at org-dartlang-sdk:///dart-sdk/lib/async/zone.dart:1795:45 trampoline @ main.dart.wasm:0x815c6a
$_Zone._processUncaughtError @ main.dart.wasm:0x815e1c
$_CustomZone.handleUncaughtError @ main.dart.wasm:0x816423
$_Future._propagateToListeners @ main.dart.wasm:0x3049ed
$_Future._completeError @ main.dart.wasm:0x305c61
$_Future._asyncCompleteError closure at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:759:29 @ main.dart.wasm:0x305eda
$closure wrapper at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:759:29 trampoline @ main.dart.wasm:0x305ef6
$_rootRun @ main.dart.wasm:0x302d48
$_rootRun tear-off trampoline @ main.dart.wasm:0x817f59
$_CustomZone.run @ main.dart.wasm:0x81659f
$_CustomZone.runGuarded @ main.dart.wasm:0x81676f
$_CustomZone.bindCallbackGuarded closure at org-dartlang-sdk:///dart-sdk/lib/async/zone.dart:1249:12 @ main.dart.wasm:0x8180b6
$closure wrapper at org-dartlang-sdk:///dart-sdk/lib/async/zone.dart:1249:12 trampoline @ main.dart.wasm:0x8180d4
$_microtaskLoop @ main.dart.wasm:0x303055
$_startMicrotaskLoop @ main.dart.wasm:0x302f83
$_startMicrotaskLoop tear-off trampoline @ main.dart.wasm:0x302fff
$_invokeCallback @ main.dart.wasm:0x2dca42
(anonymous) @ main.dart.mjs:635
load (async)
_1861 @ main.dart.mjs:662
$_EventStreamSubscription._tryResume @ main.dart.wasm:0x380a4f
$_EventStreamSubscription.onData @ main.dart.wasm:0x380873
$Stream.first @ main.dart.wasm:0x4a877e
$BrowserClient.send inner @ main.dart.wasm:0x4b81ce
$_awaitHelper closure at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:83:16 @ main.dart.wasm:0x8183af
$closure wrapper at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:83:16 trampoline @ main.dart.wasm:0x8184b5
$_rootRunUnary @ main.dart.wasm:0x3057e0
$_rootRunUnary tear-off trampoline @ main.dart.wasm:0x817ef3
$_CustomZone.runUnary @ main.dart.wasm:0x816525
$_FutureListener.handleValue @ main.dart.wasm:0x30556f
$_Future._propagateToListeners closure handleValueCallback at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:859:33 @ main.dart.wasm:0x304f15
$_Future._propagateToListeners @ main.dart.wasm:0x304b47
$_Future._completeWithValue @ main.dart.wasm:0x305c3f
$_Future._asyncCompleteWithValue closure at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:735:29 @ main.dart.wasm:0x31f404
$closure wrapper at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:735:29 trampoline @ main.dart.wasm:0x31f420
$_rootRun @ main.dart.wasm:0x302d48
$_rootRun tear-off trampoline @ main.dart.wasm:0x817f59
$_CustomZone.run @ main.dart.wasm:0x81659f
$_CustomZone.runGuarded @ main.dart.wasm:0x81676f
$_CustomZone.bindCallbackGuarded closure at org-dartlang-sdk:///dart-sdk/lib/async/zone.dart:1249:12 @ main.dart.wasm:0x8180b6
$closure wrapper at org-dartlang-sdk:///dart-sdk/lib/async/zone.dart:1249:12 trampoline @ main.dart.wasm:0x8180d4
$_microtaskLoop @ main.dart.wasm:0x303055
$_startMicrotaskLoop @ main.dart.wasm:0x302f83
$_startMicrotaskLoop tear-off trampoline @ main.dart.wasm:0x302fff
$_invokeCallback @ main.dart.wasm:0x2dca42
(anonymous) @ main.dart.mjs:635
setTimeout (async)
_1821 @ main.dart.mjs:629
$_JSEventLoop._setTimeout @ main.dart.wasm:0x30320b
$_OneShotTimer._schedule @ main.dart.wasm:0x30317d
$new _Timer (constructor body) @ main.dart.wasm:0x303138
$new _OneShotTimer (constructor body) @ main.dart.wasm:0x30311f
$_OneShotTimer @ main.dart.wasm:0x3030eb
$Timer._createTimer @ main.dart.wasm:0x30309b
$_rootCreateTimer @ main.dart.wasm:0x8173e3
$_rootCreateTimer tear-off trampoline @ main.dart.wasm:0x817450
$_CustomZone.createTimer @ main.dart.wasm:0x8162ba
$new Timer @ main.dart.wasm:0x30297e
$Timer.run @ main.dart.wasm:0x308d48
$_WidgetsFlutterBinding&BindingBase&GestureBinding&SchedulerBinding&ServicesBinding&PaintingBinding&SemanticsBinding&RendererBinding&WidgetsBinding.scheduleAttachRootWidget @ main.dart.wasm:0x307bf2
$_runWidget @ main.dart.wasm:0x307ba6
$runApp @ main.dart.wasm:0x307b28
$runDevTools closure at file:///C:/Dev/Google/devtools/packages/devtools_app/lib/initialization.dart:35:22 inner @ main.dart.wasm:0x307737
$_awaitHelper closure at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:83:16 @ main.dart.wasm:0x8183af
$closure wrapper at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:83:16 trampoline @ main.dart.wasm:0x8184b5
$_rootRunUnary @ main.dart.wasm:0x3057e0
$_rootRunUnary tear-off trampoline @ main.dart.wasm:0x817ef3
$_CustomZone.runUnary @ main.dart.wasm:0x816525
$_FutureListener.handleValue @ main.dart.wasm:0x30556f
$_Future._propagateToListeners closure handleValueCallback at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:859:33 @ main.dart.wasm:0x304f15
$_Future._propagateToListeners @ main.dart.wasm:0x304b47
$_Future._completeWithValue @ main.dart.wasm:0x305c3f
$_Future._asyncCompleteWithValue closure at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:735:29 @ main.dart.wasm:0x31f404
$closure wrapper at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:735:29 trampoline @ main.dart.wasm:0x31f420
$_rootRun @ main.dart.wasm:0x302d48
$_rootRun tear-off trampoline @ main.dart.wasm:0x817f59
$_CustomZone.run @ main.dart.wasm:0x81659f
$_CustomZone.runGuarded @ main.dart.wasm:0x81676f
$_CustomZone.bindCallbackGuarded closure at org-dartlang-sdk:///dart-sdk/lib/async/zone.dart:1249:12 @ main.dart.wasm:0x8180b6
$closure wrapper at org-dartlang-sdk:///dart-sdk/lib/async/zone.dart:1249:12 trampoline @ main.dart.wasm:0x8180d4
$_microtaskLoop @ main.dart.wasm:0x303055
$_startMicrotaskLoop @ main.dart.wasm:0x302f83
$_startMicrotaskLoop tear-off trampoline @ main.dart.wasm:0x302fff
$_invokeCallback @ main.dart.wasm:0x2dca42
(anonymous) @ main.dart.mjs:635
main.dart.mjs:56 Type 'String' is not a subtype of type 'bool' in type cast
main.dart.mjs:56     at _isFlutterGAEnabled inner (http://127.0.0.1:9101/main.dart.wasm:wasm-function[34885]:0x5b2405)
    at _awaitHelper closure at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:83:16 (http://127.0.0.1:9101/main.dart.wasm:wasm-function[62766]:0x8183af)
    at closure wrapper at org-dartlang-sdk:///dart-sdk/lib/_internal/wasm/lib/async_patch.dart:83:16 trampoline (http://127.0.0.1:9101/main.dart.wasm:wasm-function[62772]:0x8184b5)
    at _rootRunUnary (http://127.0.0.1:9101/main.dart.wasm:wasm-function[2992]:0x3057e0)
    at _rootRunUnary tear-off trampoline (http://127.0.0.1:9101/main.dart.wasm:wasm-function[62738]:0x817ef3)
    at _CustomZone.runUnary (http://127.0.0.1:9101/main.dart.wasm:wasm-function[62613]:0x816525)
    at _FutureListener.handleValue (http://127.0.0.1:9101/main.dart.wasm:wasm-function[2985]:0x30556f)
    at _Future._propagateToListeners closure handleValueCallback at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:859:33 (http://127.0.0.1:9101/main.dart.wasm:wasm-function[2963]:0x304f15)
```

**Comment by DanTup** [comment]
Since that code looks bogus to me, I retried in non-wasm mode and I see errors in the console - however unlike the wasm version, it doesn't stop DevTools loading (which is why I didn't notice it before). In wasm mode, I just get a white page with nothing rendered, but with JS DevTools loads and appears to function.

![image](https://github.com/flutter/devtools/assets/1078012/40b4d91f-8587-45c8-98e1-aaff91d12780)

(@kenzieschmoll this seems like a bug though - that API seems to be hard-coded to return an empty string and the code doesn't seem to handle empty strings, it assumes null or a boolean?)

**Comment by kenzieschmoll** [comment]
> this seems like a bug though - that API seems to be hard-coded to return an empty string and the code doesn't seem to handle empty strings, it assumes null or a boolean?).

Yes this regressed recently. Fix is here: https://github.com/flutter/devtools/pull/7896

**Comment by kenzieschmoll** [comment]
@DanTup now that the regression with accessing the Flutter store file has been fixed, are you able to verify whether the wasm build works properly when embedded in VS Code now?

**Comment by DanTup** [comment]
cross-origin-isolate is still gated with a flag in VS Code (so normal users can't get a wasm version), but if I:

1. Run `code --enable-coi` to enable cross-origin-isolate
2. Patch in ` ..add('Cross-Origin-Resource-Policy', 'cross-origin')` into the DevTools server
3. Set `customDevTools` in VS Code to run with `--wasm`

Then it _loads_:

![image](https://github.com/flutter/devtools/assets/1078012/f1670ddd-88d5-4b12-b749-10f74a54a728)

However, I can't interact with the inspector.. clicking on items in the tree just doesn't do anything (no visible selection or anything). I do now have this error in the console, though it appears as the page loads (not when I click) so I'm not sure it's related:

![image](https://github.com/flutter/devtools/assets/1078012/935fef60-d112-43fb-a4a9-081a60b94701)

However, if I open externally in the browser, then all is good (clicking in the inspector works), so it does seem like there's more broken in VS Code (I don't know if this might be why COI is still behind a flag or unrelated). I'll need to do some more debugging to try and figure out what's wrong here.

**Comment by kevmoo** [comment]
@DanTup – the cast is likely here: https://github.com/flutter/devtools/blob/86ef045a4cca71c8fdbd6a5931d80b049f93440e/packages/devtools_app/lib/src/shared/server/_release_notes_api.dart#L32

How hard is it to create a `_decodeBoyAsBool` function with some debug printing (instead of `(json.decode(respo.body) as bool)` – so we can see what's happening there?

**Comment by DanTup** [comment]
Ah, thanks for the pointer. I can just use the dev tools to see the network response - and it's a string:

![image](https://github.com/flutter/devtools/assets/1078012/41c3331b-a430-4a98-902b-5d1d2b179a3c)

Looks like this was actually fixed recently via https://github.com/flutter/devtools/pull/7997 but I didn't have a `dependency_override` in DDS to get that fix in the server.

I wasn't expecting it,  but re-running with that fix included I can use the inspector tree fine :)

![image](https://github.com/flutter/devtools/assets/1078012/34892e0d-66c7-47ab-b593-4072ec7837ab)

Maybe the release notes overlay was partially there absorbing the clicks or something? 🤷‍♂️

So, in summary:

- if the new version of devtools_shared gets rolled into the SDK
- and we add `..add('Cross-Origin-Resource-Policy', 'cross-origin')` to the DevTools server (I don't if there are security implications to consider before doing this)
- and we run VS Code with COI enabled (currently `code --enable-coi`)

Then it does appear to work (at least in my limited testing of clicking around the inspector). I don't know if/when VS Code will enable COI by default though (there's an issue at https://github.com/microsoft/vscode/issues/186614 that I think is related).

**Comment by kenzieschmoll** [comment]
> Then it does appear to work (at least in my limited testing of clicking around the inspector). I don't know if/when VS Code will enable COI by default though (there's an issue at https://github.com/microsoft/vscode/issues/186614 that I think is related).

Is there a way we can detect whether this is enabled, and fallback to using the dart2js version of DevTools if it is not?

**Comment by kevmoo** [comment]
@kenzieschmoll – we'll have to ask @eyebrowsoffire – or just try it out. I'm pretty sure we do automatically fallback in most failure cases, just not sure of this one.

**Comment by DanTup** [comment]
> Is there a way we can detect whether this is enabled, and fallback to using the dart2js version of DevTools if it is not?

This already happens, because `window.crossOriginIsolated` is false in that case and the WASM version is only selected when it's true (plus some other conditions noted further up).

**Comment by DanTup** [comment]
But, I think we do need to add that extra header in to the DevTools server, otherwise I think we could load the WASM version but then fail with the error shown above at https://github.com/flutter/devtools/issues/7856#issuecomment-2147440081.

**Comment by eyebrowsoffire** [comment]
Note: We do fall back to CanvasKit + JS if `window.crossOriginIsolated` is false

**Comment by kenzieschmoll** [comment]
This work is complete. DevTools + Wasm will be shipped as an opt-in feature behind a DevTools setting for the Q4 Dart / Flutter stable release. We will aim to turn this on by default for the Q1 2025 stable release, tracked here: https://github.com/flutter/devtools/issues/8395

**Comment by DanTup** [comment]
@kenzieschmoll FYI, I did more testing of the wasm version in VS Code this morning, and I was unable to repro the performance issues I'd recalled (which we discussed a little in https://github.com/Dart-Code/Dart-Code/pull/5367). I also tried embedding DevTools in an iframe outside of VS Code (so no Cross-Origin-Isolation) and failed to reproduce there.

I'll leave the wasm experiment on and keep an eye on it, but as things are I think the issue is gone (or, it's a much smaller issue than it was when I first hit it).

**Comment by kevmoo** [comment]
That's great new @DanTup ! Thanks for reporting!
